xla/util.h

```diff
@@ -1018,6 +1018,53 @@ struct FreeDeleter {
 std::unique_ptr<void, FreeDeleter> AlignedAlloc(std::size_t alignment,
                                                 std::size_t size);
 
+// Serialized static variable initialization from the `init_fn` output.
+// It avoids a deadlock between static variable initialization lock
+// and a lock in the `init_fn` function.
+// We have the following assumptions on `init_fn` function:
+// a) it can call python code and may release the GIL.
+// When the function is called we do not hold any non-GIL or
+// free-threading mutex.
+// b) function can be called multiple times if invoked concurrently,
+// but the output from all but one will be discarded.
+//
+// Usage:
+// Instead of incorrect code with potential deadlock
+// static SomeType* p = [](){
+//     // for example we call some python code using nanobind
+//     nb::module_ numpy = nb::module_::import_("numpy");
+//     auto np_int8 = nb::object(numpy.attr("int8"));
+//     SomeType* obj = new SomeType(np_uint8);
+//     return obj;
+// }();
+//
+// We use SafeStaticInit
+// auto func = [](){
+//     // for example we call some python code using nanobind
+//     nb::module_ numpy = nb::module_::import_("numpy");
+//     auto np_int8 = nb::object(numpy.attr("int8"));
+//     SomeType* obj = new SomeType(np_uint8);
+//     return obj;
+// }
+// SomeType* p = SafeStaticInit<SomeType>(func);
+template<typename T, typename F>
+T* SafeStaticInit(F init_fn) {
+  static absl::Mutex mutex;
+  static T* output = nullptr;
+  {
+    absl::MutexLock lock(&mutex);
+    if (output) {
+      return output;
+    }
+  }
+  std::unique_ptr<T> p(init_fn());
```

Comment:
It is safer if init_fn() allocates and returns a `std::unique_ptr` rather than you wrapping it outside the function, consider exception safety for example.